nifi-commons/nifi-processor-utilities/src/main/java/org/apache/nifi/processor/util/bin/BinFiles.java

```diff
@@ -273,25 +262,26 @@ private int binFlowFiles(final ProcessContext context, final ProcessSessionFacto
             }
 
             final ProcessSession session = sessionFactory.createSession();
-            FlowFile flowFile = session.get();
-            if (flowFile == null) {
+            final List<FlowFile> flowFiles = session.get(1000);
+            if (flowFiles.isEmpty()) {
                 break;
             }
 
-            flowFile = this.preprocessFlowFile(context, session, flowFile);
-
-            String groupId = this.getGroupId(context, flowFile);
-
-            final boolean binned = binManager.offer(groupId, flowFile, session);
-
-            // could not be added to a bin -- probably too large by itself, so create a separate bin for just this guy.
-            if (!binned) {
-                Bin bin = new Bin(0, Long.MAX_VALUE, 0, Integer.MAX_VALUE, null);
-                bin.offer(flowFile, session);
-                this.readyBins.add(bin);
+            final Map<String, List<FlowFile>> flowFileGroups = new HashMap<>();
+            for (FlowFile flowFile : flowFiles) {
+                flowFile = this.preprocessFlowFile(context, session, flowFile);
+                final String groupingIdentifier = getGroupId(context, flowFile);
+                flowFileGroups.computeIfAbsent(groupingIdentifier, id -> new ArrayList<>()).add(flowFile);
```

Comment:
I don't believe so. Using putIfAbsent, we would be creating a new ArrayList every time. By using the computeIfAbsent, it allows us to create the ArrayList only if the key is not already present.